databricks/koalas/indexing.py

```diff
@@ -400,16 +399,43 @@ def _select_rows(self, rows_sel):
                 # If slice is None - select everything, so nothing to do
                 return None, None
             elif len(self._internal.index_columns) == 1:
+                sdf = self._kdf_or_kser._internal.sdf
+                index_column = self._kdf_or_kser.index.to_series()
+                index_data_type = index_column.spark_type
                 start = rows_sel.start
                 stop = rows_sel.stop
+                order_column = sdf[NATURAL_ORDER_COLUMN_NAME]
+
+                # get natural order from '__natural_order__' from start to stop
+                # to keep natural order.
+                start_and_stop = (
+                    sdf.select(index_column._scol, NATURAL_ORDER_COLUMN_NAME)
+                       .where((index_column._scol == start) | (index_column._scol == stop))
+                       .collect())
+
+                start = [row[1] for row in start_and_stop if row[0] == start]
+                start = start[0] if len(start) > 0 else None
+
+                stop = [row[1] for row in start_and_stop if row[0] == stop]
+                stop = stop[0] if len(stop) > 0 else None
+
+                if isinstance(index_data_type, StringType):
```

Comment:
What's wrong with `StringType`?

Reply:
@ueshin thanks for the comment :)

As described at https://github.com/databricks/koalas/issues/1158#issuecomment-569570936 ,

the comparison with `StringType` couldn't keep the natural order since it literally compare the index by alphabetically. (that's why the issue #1158 is occured)

so we replace `StringType` into natural order using column `__natural_order__`.

For example, assuming that we have `sdf` like below,

```
>>> kdf._internal.sdf.show()
+-----------------+---------+------+-----------------+
|__index_level_0__|max_speed|shield|__natural_order__|
+-----------------+---------+------+-----------------+
|            cobra|        1|     2|      25769803776|
|            viper|        4|     5|      60129542144|
|       sidewinder|        7|     8|      94489280512|
+-----------------+---------+------+-----------------+
```

and if the given key is `['cobra':'viper']`, we replace it into `[25769803776:60129542144]`.

that's why we're checking and casting `StringType` into `LongType` here.

Reply:
Now we are trying to use `__natural_order__` to calculate actual start and stop regardless of the data type. I don't think we consider the data type here again. WDYT?

Reply:
@ueshin 

Ah, i got what you meant.

yes i totally agree with you and this part should be corrected.

but FYI, i finally remember that the very first reason why i check the `StringType` here.

at the first time, i thought the behavior of `loc` is slightly different between when `StringType` and `LongType` because of the below example.

```python
>>> pdf1
   max_speed  shield
A          1       2
C          4       5
B          7       8

>>> pdf2
    max_speed  shield
10          1       2
20          4       5
30          7       8

>>> pdf1.loc['A':'Z']
Traceback (most recent call last):
...
ValueError: index must be monotonic increasing or decreasing

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
...
KeyError: 'Z'

>>> pdf2.loc[10:10000]
    max_speed  shield
10          1       2
20          4       5
30          7       8
```

if the given key of slice, `start` or `stop`, is not included list of index, i thought that the `StringType` raise `KeyError` while `LongType` works well, but it was wrong.

this different was not come from the type of index, but `ValueError: index must be monotonic increasing or decreasing`.

And the cases that index has not index as monotonic increasing or decreasing order is missing our test cases for `loc`, i had have mistaken.

i'll make it work properly soon.

thanks for the comment! :)